python-frontend/src/main/java/org/sonar/python/semantic/FunctionSymbolImpl.java

```diff
@@ -218,9 +227,28 @@ private void addParameter(org.sonar.plugins.python.api.tree.Parameter parameter,
     Name parameterName = parameter.name();
     Token starToken = parameter.starToken();
     if (parameterName != null) {
-      InferredType declaredType = getParameterType(parameter, starToken);
+      InferredType declaredType = null;
+      boolean isPositionalVariadic = false;
+      boolean isKeywordVariadic = false;
+      if (starToken != null) {
+        // https://docs.python.org/3/reference/compound_stmts.html#function-definitions
+        hasVariadicParameter = true;
+        if ("*".equals(starToken.value())) {
+          // if the form “*identifier” is present, it is initialized to a tuple receiving any excess positional parameters
+          isPositionalVariadic = true;
+          declaredType = InferredTypes.TUPLE;
+        }
+        if ("**".equals(starToken.value())) {
+          //  If the form “**identifier” is present, it is initialized to a new ordered mapping receiving any excess keyword arguments
+          isKeywordVariadic = true;
+          declaredType = InferredTypes.DICT;
+        }
+      }
```

Comment:
can we keep this inside `getParameterType` method? I think `addParameter` is already quite long and it has Cognitive complexity of 18

Reply:
The issue is that this piece of code now produces the declared type as well as the boolean `isPositionalVariadic` and `isKeywordVariadic`.

While it's a bit complex due to the nesting and the multiple if/else, I felt it still was more understandable that way than if I wrapped this info in a different structure for example.